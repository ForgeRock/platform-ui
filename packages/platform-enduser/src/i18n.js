/**
 * Copyright (c) 2020-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { createI18n } from 'vue-i18n';
import { merge } from 'lodash';
import dateTimeFormats from '@forgerock/platform-shared/src/dateTimeFormats.json';

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  // Can't use workspace symlink due to jest issues
  const sharedLocales = require.context('../../platform-shared/src/locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      let generatedTranslation = locales(key);

      // Check for shared local and merge in if it exists
      const sharedLocaleExists = sharedLocales.keys().includes(key);
      if (sharedLocaleExists) {
        generatedTranslation = merge(sharedLocales(key), generatedTranslation);
      }

      messages[locale] = generatedTranslation;
    }
  });

  return messages;
}

export default createI18n({
  messages: loadLocaleMessages(),
  silentFallbackWarn: true,
  silentTranslationWarn: true,
  dateTimeFormats,
});
