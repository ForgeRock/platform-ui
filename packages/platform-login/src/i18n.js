/**
 * Copyright (c) 2020-2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Vue from 'vue';
import VueI18n from 'vue-i18n';
import { merge } from 'lodash';
import dateTimeFormats from '@forgerock/platform-shared/src/dateTimeFormats.json';

Vue.use(VueI18n);

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  // Can't use workspace symlink due to jest issues
  const sharedLocales = require.context('../../platform-shared/src/locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      let generatedTranslation = locales(key);

      // Check for shared local and merge in if it exists
      const sharedLocaleExists = sharedLocales.keys().includes(key);
      if (sharedLocaleExists) {
        generatedTranslation = merge(sharedLocales(key), generatedTranslation);
      }

      messages[locale] = generatedTranslation;
    }
  });

  return messages;
}

Vue.mixin({
  methods: {
    translationExists(path) {
      return this.$t(path) !== path;
    },
  },
});

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages(),
  silentTranslationWarn: true,
  dateTimeFormats,
});
