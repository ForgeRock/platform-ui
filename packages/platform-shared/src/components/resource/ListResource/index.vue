<!-- Copyright 2019-2020 ForgeRock AS. All Rights Reserved

Use of this code requires a commercial software license with ForgeRock AS.
or with one of its affiliates. All use shall be exclusively subject
to such license between the licensee and ForgeRock AS. -->
<template>
  <div>
    <div class="card">
      <div class="card-header py-2">
        <BRow>
          <BCol
            md="7"
            class="my-1">
            <slot
              name="listToolbar" />
          </BCol>
          <BCol
            md="5"
            class="my-1">
            <FrSearchInput
              @search="search"
              @clear="clear"
              :placeholder="$t('common.search')"
              v-model="filter"
            />
          </BCol>
        </BRow>
      </div>
      <BTable
        show-empty
        table-responsive
        stacked="lg"
        :items="tableData"
        :fields="columns"
        :per-page="0"
        :hover="tableHover"
        :sort-by.sync="sortBy"
        :sort-desc.sync="sortDesc"
        :no-local-sorting="true"
        class="mb-0"
        sort-direction="asc"
        @row-clicked="$emit('rowClicked', $event)"
        @sort-changed="sortingChanged">
        <template v-slot:cell(actions)="data">
          <slot
            name="actions"
            :item="data">
            <BDropdown
              variant="link"
              no-caret
              right
              toggle-class="text-decoration-none p-0">
              <template v-slot:button-content>
                <i class="material-icons-outlined text-muted md-24">
                  more_horiz
                </i>
              </template>
              <BDropdownItem @click="editResource(data.item)">
                <i class="material-icons-outlined mr-3">
                  edit
                </i> {{ $t('common.edit') }}
              </BDropdownItem>
              <BDropdownItem
                @click="confirmDeleteResource(data.item._id)">
                <i class="material-icons-outlined mr-3">
                  delete
                </i> {{ $t('common.delete') }}
              </BDropdownItem>
            </BDropdown>
          </slot>
        </template>
      </BTable>
      <div class="card-footer py-2">
        <nav aria-label="Page navigation">
          <ul class="pagination justify-content-center mb-0">
            <li
              @click.prevent="currentPage === 1 ? '' : paginationChange(1)"
              :class="[{ disabled: currentPage === 1 }, 'page-item']">
              <a
                class="page-link"
                href="#">
                <i class="material-icons material-icons-outlined mr-2">
                  first_page
                </i>
              </a>
            </li>
            <li
              @click.prevent="currentPage === 1 ? '' : paginationChange(currentPage - 1)"
              :class="[{ disabled: currentPage === 1 }, 'page-item']">
              <a
                class="page-link"
                href="#">
                <i class="material-icons material-icons-outlined mr-2">
                  keyboard_arrow_left
                </i>
              </a>
            </li>
            <li
              @click.prevent="lastPage ? '' : paginationChange(currentPage + 1)"
              :class="[{ disabled: lastPage }, 'page-item']">
              <a
                class="page-link"
                href="#">
                <i class="material-icons material-icons-outlined mr-2">
                  keyboard_arrow_right
                </i>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>

    <slot name="deleteResourceModal">
      <FrDeleteResource
        modal-id="deleteResource"
        @resource-deleted="loadData('true', displayFields, null, 1)"
        @cancel-resource-deletion="clearSelection()"
        :resource-name="resourceName"
        :resource-type="resourceType"
        :resource-to-delete-id="resourceToDeleteId"
        :delete-managed-resource="deleteManagedResource"
        :delete-internal-resource="deleteInternalResource" />
    </slot>
  </div>
</template>

<script>
import {
  includes,
  isNull,
  each,
  isUndefined,
  isNaN,
  toNumber,
} from 'lodash';
import {
  BCol,
  BDropdown,
  BDropdownItem,
  BRow,
  BTable,
  VBModal,
} from 'bootstrap-vue';
import Vue from 'vue';
import NotificationMixin from '@forgerock/platform-shared/src/mixins/NotificationMixin';
import SearchInput from '@forgerock/platform-shared/src/components/SearchInput';
import DeleteResource from '../DeleteResource';

Vue.directive('b-modal', VBModal);
/**
 * @description Controlling component for delegated admin, allows for listing available resources and connects to the create, delete and edit features.
 *
 * @fires GET resourceType/resourceName?_queryFilter=filter&_pageSize=10&_totalPagedResultsPolicy=EXACT (e.g. managed/user?_queryFilter=true&_pageSize=10&_totalPagedResultsPolicy=EXACT) -
 * List resource items, limited to 10 returned items and makes use of a query filter search if provided (defaults to queryFilter = true if none provided by the user).
 */
export default {
  name: 'ListResource',
  mixins: [
    NotificationMixin,
  ],
  components: {
    BCol,
    BDropdown,
    BDropdownItem,
    BRow,
    BTable,
    FrDeleteResource: DeleteResource,
    FrSearchInput: SearchInput,
  },
  directives: {
    'b-modal': VBModal,
  },
  props: {
    deleteManagedResource: {
      type: Function,
      default: () => {
        const retv = {
          then: () => {},
        };
        return retv;
      },
    },
    deleteInternalResource: {
      type: Function,
      default: () => {
        const retv = {
          then: () => {},
        };
        return retv;
      },
    },
    routerParameters: {
      type: Object,
      default: () => {},
    },
    tableData: {
      type: Array,
      default: () => [],
    },
    lastPage: {
      type: Boolean,
      default: true,
    },
  },
  data() {
    return {
      resourceName: this.$route.params.resourceName,
      resourceType: this.$route.params.resourceType,
      isRowSelected: false,
      tableHover: true,
      columns: [],
      displayFields: [],
      currentPage: 1,
      sortBy: null,
      sortDesc: false,
      filter: '',
      resourceToDeleteId: '',
    };
  },
  mounted() {
    this.resourceName = this.getResourceName(this.routerParameters.resourceName);
    this.loadTableDefs();
    this.loadData('true', this.displayFields, '', 1);
  },
  methods: {
    getResourceName(resourceName) {
      if (isUndefined(resourceName)) {
        return this.$route.params.resourceName;
      }
      return resourceName;
    },
    /**
     * Sets up sorting for API call URL, considering ascending/descending
     *
     * @param {boolean} sortDesc - Required toggle on if we are sorting in descending order
     * @param {string} sortBy - Required name of field to sort table on
     */
    calculateSort(sortDesc, sortBy) {
      let sortUrl = null;

      if (!isNull(sortBy)) {
        if (sortDesc) {
          sortUrl = `${sortBy}`;
        } else {
          sortUrl = `-${sortBy}`;
        }
      }
      return sortUrl;
    },
    /**
     * Reloads data after search box filter text is cleared
     */
    clear() {
      this.filter = '';
      this.sortBy = null;
      this.sortDesc = false;
      this.currentPage = 1;

      this.loadData('true', this.displayFields, null, 1);
    },
    /**
     * Builds API URL using value in search box
     *
     * @param {string} filter - Required current value of search box
     * @param {array} displayFields - Required array of field names that we want to query on
     * @param {object} schemaProps - Required metadata of current schema
     */
    generateSearch(filter, displayFields, schemaProps) {
      let filterUrl = '';

      if (filter.length > 0) {
        each(displayFields, (field, index) => {
          let type = 'string';

          if (!isUndefined(schemaProps) && !isUndefined(schemaProps[field])) {
            // eslint-disable-next-line prefer-destructuring
            type = schemaProps[field].type;
          }

          if (type === 'number' && !isNaN(toNumber(filter))) {
            // Search based on number and proper number value
            if ((index + 1) < displayFields.length) {
              filterUrl = `${filterUrl}${field}+eq+ ${filter}+OR+`;
            } else {
              filterUrl = `${filterUrl}${field}+eq+ ${filter}`;
            }
          } else if (type === 'boolean' && (filter === 'true' || filter === 'false')) {
            // Search based on boolean and proper boolean true/false
            if ((index + 1) < displayFields.length) {
              filterUrl = `${filterUrl}${field}+eq+ ${filter}+OR+`;
            } else {
              filterUrl = `${filterUrl}${field}+eq+ ${filter}`;
            }
          } else if ((index + 1) < displayFields.length) {
            // Fallback to general string search if all other criteria fails
            filterUrl = `${filterUrl}${field}+sw+"${filter}"+OR+`;
          } else {
            filterUrl = `${filterUrl}${field}+sw+"${filter}"`;
          }
        });
      } else {
        filterUrl = 'true';
      }

      return filterUrl;
    },
    loadData(filter, fields, sortField, page) {
      this.$emit('getTableData', {
        filter,
        fields,
        sortField,
        page,
      });
    },
    /**
     * Builds table column definitions using menu item parameters
     */
    loadTableDefs() {
      this.displayFields = [];
      this.columns = [];
      if (this.routerParameters && this.routerParameters.order) {
        this.routerParameters.order.forEach((columnName) => {
          const column = this.routerParameters.managedProperties[columnName];
          if (includes(['string', 'boolean', 'number'], column.type)) {
            if (this.displayFields.length < 4 && column.viewable) {
              this.displayFields.push(columnName);
              this.columns.push({
                key: columnName,
                label: column.title,
                sortable: column.sortable,
                sortDirection: 'desc',
              });
            }
          }
        });

        /*
         * Push a final column for the "actions" menu.
         * Empty label is intended.
         */
        this.columns.push({
          key: 'actions',
          label: '',
        });
      }
    },
    /**
     * Repulls data based on new table page
     *
     * @param {number} page - Required integer number specifying which table page we are viewing
     */
    paginationChange(page) {
      this.currentPage = page;
      this.loadData(this.generateSearch(this.filter, this.displayFields, this.routerParameters.managedProperties), this.displayFields, this.calculateSort(this.sortDesc, this.sortBy), page);
    },
    clearSelection() {
      this.resourceToDeleteId = '';
    },
    /**
     * Attempts to pull up edit resource page for selected row if user has access
     *
     * @param {object} item - Required metaData of row clicked
     */
    editResource(item) {
      if (this.userCanUpdate) {
        this.$router.push({
          name: 'EditResource',
          params: {
            resourceName: this.resourceName,
            resourceType: this.resourceType,
            // eslint-disable-next-line no-underscore-dangle
            resourceId: item._id,
          },
        });
      } else {
        this.displayNotification('IDMMessages', 'error', this.$t('userflow.unableToEditResource', { resource: this.resourceName }));
      }
    },
    confirmDeleteResource(id) {
      this.resourceToDeleteId = id;
    },
    /**
     * Repulls data based on input search box text
     */
    search() {
      if (this.filter === '') {
        this.clear();
        return;
      }
      this.sortBy = null;
      this.sortDesc = false;
      this.currentPage = 1;

      this.loadData(this.generateSearch(this.filter, this.displayFields, this.routerParameters.managedProperties), this.displayFields, null, 1);
    },
    /**
     * Repulls data based on new sort, and returns table to first page
     *
     * @param {object} sort - Required object containing sort metadata
     */
    sortingChanged(sort) {
      this.currentPage = 1;

      this.loadData(this.generateSearch(this.filter, this.displayFields, this.routerParameters.managedProperties), this.displayFields, this.calculateSort(sort.sortDesc, sort.sortBy), 1);
    },
  },
};
</script>
