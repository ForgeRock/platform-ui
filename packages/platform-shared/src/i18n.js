/**
 * Copyright 2019 ForgeRock AS. All Rights Reserved
 *
 * Use of this code requires a commercial software license with ForgeRock AS.
 * or with one of its affiliates. All use shall be exclusively subject
 * to such license between the licensee and ForgeRock AS.
 */
import Vue from 'vue';
import VueI18n from 'vue-i18n';
import dateTimeFormats from './dateTimeFormats.json';

Vue.use(VueI18n);

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      messages[locale] = locales(key);
    }
  });

  return messages;
}

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages(),
  silentTranslationWarn: true,
  dateTimeFormats,
});
