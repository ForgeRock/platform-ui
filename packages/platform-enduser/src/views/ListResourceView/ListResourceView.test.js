/**
 * @license
 * Copyright (c) 2020 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import BootstrapVue from 'bootstrap-vue';
import flushPromises from 'flush-promises';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import i18n from '@/i18n';
import ListResource from './index';

const localVue = createLocalVue();
localVue.use(BootstrapVue);
ListResource.mounted = jest.fn();
function flush() {
  flushPromises();
  jest.runAllTimers();
}

describe('ListResource.vue', () => {
  let wrapper;

  beforeEach(() => {
    jest.useFakeTimers();
    const $route = {
      path: '/test',
      meta: {},
      params: {
        resourceName: 'test',
        resourceType: 'test',
      },
    };

    wrapper = shallowMount(ListResource, {
      localVue,
      i18n,
      stubs: {
        'router-link': true,
      },
      mocks: {
        $route,
      },
    });
    wrapper.setMethods({ loadGrid: () => { } });
  });

  it('ListResourceView page loaded', () => {
    expect(wrapper.name()).toBe('ListResourceView');
    expect(wrapper).toMatchSnapshot();
  });

  it('Builds URL Parameters', () => {
    expect(wrapper.vm.buildUrlParams('name+sw+"test"+OR+description+sw+"test"', ['name', 'description'], '', 1)).toEqual({
      fields: 'name,description',
      pageSize: 10,
      queryFilter: 'name+sw+"test"+OR+description+sw+"test"',
      sortKeys: 'name',
      totalPagedResultsPolicy: 'EXACT',
    });
  });

  it('Gets Table Data', async () => {
    expect(wrapper.vm.tableData).toStrictEqual([]);
    wrapper.vm.routerParameters = { resourceType: 'managed' };
    wrapper.vm.getTableData({
      fields: ['name', 'description'],
      filter: 'name+sw+"test"+OR+description+sw+"test"',
      page: 1,
      sortField: '',
    });
    await flush();

    expect(wrapper.vm.tableData).toStrictEqual([]);
  });
});
