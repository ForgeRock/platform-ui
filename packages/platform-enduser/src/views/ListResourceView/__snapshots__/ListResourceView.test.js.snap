// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListResource.vue ListResourceView page loaded 1`] = `
<bcontainer-stub tag="div">
  <div class="mt-5">
    <frheader-stub title="Test" subtitle="Hosted identity resources"></frheader-stub>
    <bcard-stub bodytag="div" titletag="h4" subtitletag="h6" subtitletextvariant="muted" headertag="div" footertag="div" tag="div" nobody="true" class="card-tabs-vertical">
      <!---->
      <!---->
    </bcard-stub>
  </div>
</bcontainer-stub>
`;
